# Рынок видеоигр

- Автор: Авдеев Николай
- Дата: 21.12.2025

### Цели и задачи проекта

<font color='#0000000'>Целями проекта являются:
* Привлечь внимание людей, которые любят старые игры.
* Привлечение новой аудитории игрой "Секреты Темнолесья".

Задачи проекта:
* Изучение развития игровой индустрии с 2000 по 2013 год.
* Рассмотрение топ-платформ по количеству выпущенных игр за актуальный период.</font>

### Описание данных

<font color='#000000'>В проекте будут использованы данные датасета /datasets/new_games.csv с таким описанием:

* `Name` — название игры.
* `Platform` — название платформы.
* `Year of Release` — год выпуска игры.
* `Genre` — жанр игры.
* `NA sales` — продажи в Северной Америке (в миллионах проданных копий).
* `EU sales` — продажи в Европе (в миллионах проданных копий).
* `JP sales` — продажи в Японии (в миллионах проданных копий).
* `ther sales` — продажи в других странах (в миллионах проданных копий).
* `Critic Score` — оценка критиков (от 0 до 100).
* `User Score` — оценка пользователей (от 0 до 10).
* `Rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.</font>

### Содержимое проекта

1. Загрузка и знакомство с данными  
2. Проверка ошибок в данных и их предобработка:  
* Названия, или метки, столбцов датафрейма  
* Типы данных  
* Наличие пропусков в данных  
* Явные и неявные дубликаты в данных  
3. Фильтрация данных  
4. Категоризация данных

---

## 1.Загрузка и знакомство с данными:

- Загрузим необходимые библиотеки для анализа данных и данные из датасета `/datasets/new_games.csv`. Затем выведем основную информацию о данных с помощью метода info() и первые строки датафрейма.


In [1]:
#Импортируем библиотеку pandas
import pandas as pd

C:\Users\PC\Jupyter\lib\site-packages\pandas\core\computation\expressions.py:21: UserWarning: Pandas requires version '2.8.4' or newer of 'numexpr' (version '2.7.3' currently installed).
  from pandas.core.computation.check import NUMEXPR_INSTALLED
C:\Users\PC\Jupyter\lib\site-packages\pandas\core\arrays\masked.py:61: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.2' currently installed).
  from pandas.core import (


In [2]:
# Выгружаем данные из датасета /datasets/new_games.csv в датафрейм ng
ng = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')

In [3]:
# Выводим информацию о датафрейме
ng.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [4]:
# Выводим первые строки датафрейма на экран
ng.head()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


#### Датасет `/datasets/new_games.csv` содержит 11 столбцов и 16956 строк, в которых представлена информация о играх.
Всего датасет представлен в 2-х типах данных, рассмотрим их:
- Строковые данные (object) -  7 стобцов имеют тип данных `object`:
  - `Name` - содержат строковые данные (название игр), что верно для текстовой информации.
  - `Platform`, `Genre` и `Rating` - также содержат строковые данные (платформу, жанр и возрастной рейтинг), но информацию из данных столбцов можно классифицировать по группам, поэтому лучше привести тип данных в формат category.
  - `EU sales` и `JP sales` имеют данные о кол-ве проданных копий игр в Европе и Японии, информацию в данных столбцах удобнее хранить в числовых значениях с плавающей запятой (float64), т.к. может потребоваться выполнение математических операций с этими данными.
  - `User Score` содержит информацию об оценке пользователей, в этом случае также лучше изменить тип данных на float64 для выполнения операций с числовыми данными.
- Числовые значения с плавающей запятой (float64) - представленный тип данных имеют 4 столбца:
  - `Year of Release` - данный столбец хранит только год выпуска игры, т.к. в данных не представлены месяц и число, то преобразовывать тип данных в datetime64 не требуется, но для удобства работы в дальнейшем анализе, изменим тип на Int64 (т.к. столбец нельзя преобразовывать к типу int64 пока в нем числятся пропуски, а Int64 дает такую возможность). 
  - `NA sales`, `Other sales` и `Critic Score` - здесь тип данных представлен верно, в виду того, что столбцы отображает кол-во проданных копий игры в Северной Америке, в других странах и содержат оценку критиков по 100 - бальной шкале соответственно.


---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

In [5]:
#Выведем на экран название всех столбцов
ng.columns
display(f'Названия столбцов: {ng.columns}')

"Названия столбцов: Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',\n       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],\n      dtype='object')"

In [6]:
#Приведем название столбцов к стилю snake_case и нижнему регистру
ng.columns = ng.columns.str.replace(' ', '_').str.lower()
display(f'Обновленные названия столбцов: {ng.columns}')

"Обновленные названия столбцов: Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',\n       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],\n      dtype='object')"

После преобразования столбцов к стилю snake_case и нижнему регистру датасет приобретает такой вид:

- `name` — название игры.
- `platform` — название платформы.
- `year_of_release` — год выпуска игры.
- `genre` — жанр игры.
- `na_sales` — продажи в Северной Америке (в миллионах проданных копий).
- `eu_sales` — продажи в Европе (в миллионах проданных копий).
- `jp_sales` — продажи в Японии (в миллионах проданных копий).
- `other_sales` — продажи в других странах (в миллионах проданных копий).
- `critic_score` — оценка критиков (от 0 до 100).
- `user_score` — оценка пользователей (от 0 до 10).
- `rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

### 2.2. Типы данных

Причинами некорректных типов данных в нашем датасете могут быть следующими:

Столбцы `Platform`, `Genre`, `Rating`

- Импорт данных из CSV или другого источника часто приводит к тому, что все значения в столбцах интерпретируются неправильно

- Отсутствие явного указания типов при чтении файла

Столбцы`EU sales` и `JP sales`

- Возможны пропущенные значения, записанные как NaN, что может привести к типу object после некорректных чтений

- Значения в этих столбцах могли содержать текстовые символы или запятые вместо точек, вызывая ошибки при автоматическом определении типа

Столбец `User Score`

- Возможно, при чтении файла не были правильно обработаны отсутствующие или нечисловые значения

Столбец `Year of Release`

- Могут быть пропущенные значения, неправильные форматы или строки, что мешает автоматическому определению типа

In [7]:
#Преобразуем столбцы `eu_sales`, `jp_sales` и `user_score` к типу данных float64 
ng['eu_sales'] = pd.to_numeric(ng['eu_sales'], errors='coerce')
ng['jp_sales'] = pd.to_numeric(ng['jp_sales'], errors='coerce')
ng['user_score'] = pd.to_numeric(ng['user_score'], errors='coerce')

In [8]:
#Преобразуем столбцы `platform`, `genre` и `rating` к типу данных category
types_to_convert = ['platform', 'genre',  'rating']
ng[types_to_convert] = ng[types_to_convert].astype('category')

In [9]:
#Чтобы избавиться от дробной части, преобразуем столбец `year_of_release` к типу данных Int64
ng['year_of_release'] = ng['year_of_release'].astype('Int64')

In [10]:
#Проверим корректность преобразования столбцов в новые типы данных
ng[['eu_sales', 'jp_sales', 'user_score', 'year_of_release', 'platform', 'genre',  'rating']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype   
---  ------           --------------  -----   
 0   eu_sales         16950 non-null  float64 
 1   jp_sales         16952 non-null  float64 
 2   user_score       7688 non-null   float64 
 3   year_of_release  16681 non-null  Int64   
 4   platform         16956 non-null  category
 5   genre            16954 non-null  category
 6   rating           10085 non-null  category
dtypes: Int64(1), category(3), float64(3)
memory usage: 598.6 KB


### 2.3. Наличие пропусков в данных

- Посчитаем количество пропусков в каждом столбце в абсолютных и относительных значениях.


In [11]:
# Выводим количество пропущенных строк в датафрейме
ng.isna().sum()

name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8714
user_score         9268
rating             6871
dtype: int64

In [12]:
# Подсчитываем процент строк с пропусками
ng.isna().sum() / len(ng) * 100 

name                0.011795
platform            0.000000
year_of_release     1.621845
genre               0.011795
na_sales            0.000000
eu_sales            0.035386
jp_sales            0.023590
other_sales         0.000000
critic_score       51.391838
user_score         54.659118
rating             40.522529
dtype: float64

- В датасете имеются пропуски в столбцах: 
* `name` - 2 пропуска (менее 0.1%).
  Причины: Возможно записи получены без названия игры, либо это связано с некорректным сбором данных или ошибками при вводе.
  Решение: Так как пропусков крайне мало, то можно оставить их как есть или удалить.
* `year_of_release` 275 пропусков (~1.6%).
  Причины: Дата релиза неизвестна или не была зарегистрирована при выпуске игры.
  Решение: Ввиду того, что данный столбец потребуется в дальнейшем анализе, то целесообразнее будет удалить строки с пропусками для более точных результатов.
* `genre` - 2 пропуска (менее 0.1%).
  Причины: Не указана категория жанра для некоторых игр или произошла ошибка при загрузке данных.
  Решение: Аналогично `Name`, можно оставить как есть или удалить эти строки, т.к. их кол-во незначительно.
* `eu_sales` и `jp_sales` - 6 пропусков (менее 0.1%) и 4 пропуска (менее 0.1%) соответственно.
  Причины: Игры могли не продаваться в Европе или Японии, или продажи там были настолько малы, что не учитывались и не регистрировались, также в процессе сбора информации могли возникнуть технические или человеческие ошибки.
  Решение: Пропуски можно заменить на среднее значение в зависимости от названия платформы и года выхода игры, либо удалить, т.к. их кол-во незначительно.
* `critic_score` - 8714 пропусков (~51%).
  Причины: Отсутствие оценки критиков для большинства игр.
  Решение: Так как данные потребуются для дальнейшего анализа, лучше всего заменить пропуски на метку.
* `user_score` - 6804 пропусков (~40%).
  Причины: Пользователи не выставили оценку.
  Решение: Аналогично `Critic Score`, логичнее всего заменить пропуски на метку, т.к. информация будет использована для анализа.
* `rating` - 6871 пропусков (~41%).
  Причины: Вероятнее всего данные игры не были рейтированы.
  Решение: Так как данный столбец не потребуется для дальнейшего анализа, то можно оставить пропуски как есть.

- Обработаем пропущенные значения. Для каждого случая выберем оптимальный вариант.

In [13]:
#Сохраним изначальное кол-во строк в датафрейме и удалим строки с пропусками в столбцах `name`, `genre`, `year_of_release`
initial_row_count1 = ng.shape[0]
ng = ng.dropna(subset=['name', 'genre', 'year_of_release'])

In [14]:
#В столбцах `eu_sales` и `jp_sales` заменим пропуски на среднее значение в зависимости от названия платформы и года выхода игры
def mean_group_sales(row):
    if pd.isna(row[['eu_sales', 'jp_sales']]).any():
        group = ng[
            (ng['platform'] == row['platform']) &
            (ng['year_of_release'] == row['year_of_release'])
        ]
        mean_eu = group['eu_sales'].mean()
        mean_jp = group['jp_sales'].mean()
        if pd.isna(row['eu_sales']):
            row['eu_sales'] = mean_eu
        if pd.isna(row['jp_sales']):
            row['jp_sales'] = mean_jp
    return row

ng = ng.apply(mean_group_sales, axis=1)

In [15]:
#В столбцах `critic_score` и `user_score` заменим пропуски на значение-индикатор (-1)
ng[['critic_score', 'user_score']] = ng[['critic_score', 'user_score']].fillna(-1)

In [16]:
#Проверим кол-во строк после обработки пропусков
final_row_count = ng.shape[0]
display(f"Количество строк после обработки пропусков: {final_row_count}")

'Количество строк после обработки пропусков: 16679'

In [17]:
#Проверим столбцы на оставшиеся пропуски
display(ng.isna().sum())

name                  0
platform              0
year_of_release       0
genre                 0
na_sales              0
eu_sales              0
jp_sales              0
other_sales           0
critic_score          0
user_score            0
rating             6778
dtype: int64

- Исходя из данных, видим, что замена и удаление пропусков прошло успешно. Остались только пропущенные знаяения
в столбце `rating`, но т.к. данный столбец не входит анализ, то было принято решение оставить его без изменений.

In [18]:
#Выведем обновленный датафрейм
display(ng.head())

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
0,Wii Sports,Wii,2006,Sports,41.36,28.96,3.77,8.45,76.0,8.0,E
1,Super Mario Bros.,NES,1985,Platform,29.08,3.58,6.81,0.77,-1.0,-1.0,NaN
2,Mario Kart Wii,Wii,2008,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009,Sports,15.61,10.93,3.28,2.95,80.0,8.0,E
4,Pokemon Red/Pokemon Blue,GB,1996,Role-Playing,11.27,8.89,10.22,1.00,-1.0,-1.0,NaN


### 2.4. Явные и неявные дубликаты в данных

- Изучим уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверим, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведем нормализацию данных с текстовыми значениями.

In [19]:
#Проверим столбец `genre` на неявные дубликаты
ng_genre_unique = ng['genre'].unique()
display(ng_genre_unique)

array(['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', 'Misc',
       'Shooter', 'Simulation', 'Action', 'Fighting', 'Adventure',
       'Strategy', 'MISC', 'ROLE-PLAYING', 'RACING', 'ACTION', 'SHOOTER',
       'FIGHTING', 'SPORTS', 'PLATFORM', 'ADVENTURE', 'SIMULATION',
       'PUZZLE', 'STRATEGY'], dtype=object)

In [20]:
#Проверим столбец `platform` на неявные дубликаты
ng_platform_unique = ng['platform'].unique()
display(ng_platform_unique)

array(['Wii', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       'PS4', '3DS', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XOne',
       'WiiU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

In [21]:
#Проверим столбец `rating` на неявные дубликаты
ng_rating_unique = ng['rating'].unique()
display(ng_rating_unique)

array(['E', nan, 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'], dtype=object)

In [22]:
#Проверим столбец `year_of_release` на неявные дубликаты
ng_year_of_release_unique = ng['year_of_release'].unique()
display(ng_year_of_release_unique)

array([2006, 1985, 2008, 2009, 1996, 1989, 1984, 2005, 1999, 2007, 2010,
       2013, 2004, 1990, 1988, 2002, 2001, 2011, 1998, 2015, 2012, 2014,
       1992, 1997, 1993, 1994, 1982, 2016, 2003, 1986, 2000, 1995, 1991,
       1981, 1987, 1980, 1983], dtype=int64)

- Исходя из произведенного анализа на уникальные значения в столбцах `genre`, `platform`, `rating` и `year_of_release` можем сделать вывод:
  - `genre` - имеет неявные дубликаты, а именно некоторые из одних и тех же жанров игр написаны в разных регистрах. Для дальнейшей работы с данным столбцом преобразуем категории в нижний регистр.
  - `platform` - не имеет неявных дубликатов.
  - `rating` - неявных дубликатов не содержит, но имеется 1 лишний возрастной рейтинг K-A. K-A - это устаревшая категория рейтинга E. Произведем замену рейтинга K-A на E во всем столбце.
  - `year_of_release` - также не имеет неявных дубликатов.

In [23]:
#Произведем замену возрастного рейтинга K-A на E
ng['rating'] = ng['rating'].replace({'K-A': 'E'})
display(ng['rating'])

0          E
1        NaN
2          E
3          E
4        NaN
        ... 
16951    NaN
16952    NaN
16953    NaN
16954    NaN
16955    NaN
Name: rating, Length: 16679, dtype: object

In [24]:
#Преобразуем столбец в строковый тип данных для изменения названия жанров игр в нижний регитср и вернем категориальный тип
#данных
ng['genre'] = ng['genre'].astype(str).str.lower().astype('category')
display(ng['genre'])

0              sports
1            platform
2              racing
3              sports
4        role-playing
             ...     
16951          action
16952          sports
16953       adventure
16954        platform
16955      simulation
Name: genre, Length: 16679, dtype: category
Categories (12, object): ['action', 'adventure', 'fighting', 'misc', ..., 'shooter', 'simulation', 'sports', 'strategy']

- После нормализации данных и устрания неявные дубликатов, проверим наличие явных дубликатов в данных.

In [25]:
#Посчитаем кол-во явных дубликатов в датасете
ng_dublicates = ng.duplicated(keep='first').sum()
display(f"Кол-во дубликатов в датасете: {ng_dublicates}")

'Кол-во дубликатов в датасете: 235'

In [26]:
ng_sorted = ng.sort_values(by=ng.columns.tolist())
duplicates = ng_sorted[ng_sorted.duplicated(keep=False)]
display(duplicates)

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
15191,Beyblade Burst,3DS,2016,role-playing,0.00,0.00,0.03,0.00,-1.0,-1.0,NaN
15192,Beyblade Burst,3DS,2016,role-playing,0.00,0.00,0.03,0.00,-1.0,-1.0,NaN
15301,11eyes: CrossOver,X360,2009,adventure,0.00,0.00,0.02,0.00,-1.0,-1.0,NaN
15302,11eyes: CrossOver,X360,2009,adventure,0.00,0.00,0.02,0.00,-1.0,-1.0,NaN
4860,18 Wheeler: American Pro Trucker,PS2,2001,racing,0.20,0.15,0.00,0.05,61.0,5.7,E
...,...,...,...,...,...,...,...,...,...,...,...
2909,Yu-Gi-Oh! The Falsebound Kingdom,GC,2002,strategy,0.49,0.13,0.07,0.02,-1.0,-1.0,NaN
6695,Zoo Resort 3D,3DS,2011,simulation,0.11,0.09,0.03,0.02,-1.0,-1.0,E
6696,Zoo Resort 3D,3DS,2011,simulation,0.11,0.09,0.03,0.02,-1.0,-1.0,E
8156,Zumba Fitness Rush,X360,2012,sports,0.00,0.16,0.00,0.02,73.0,6.2,E10+


In [27]:
initial_row_count2 = ng.shape[0]
ng_sorted = ng.sort_values(by=list(ng.columns))
ng = ng_sorted.drop_duplicates()
final_row_count2 = ng.shape[0]
display(f"Количество строк до удаления дубликатов: {initial_row_count2}")
display(f"Количество строк после удаления дубликатов: {final_row_count2}") 

'Количество строк до удаления дубликатов: 16679'

'Количество строк после удаления дубликатов: 16444'

- В процессе анализа данных было обнаружено 235 явных дубликатов строк. Для их обработки выполнены следующие действия: данные были отсортированы по всем столбцам для группировки дубликатов, после чего все дублирующие записи были выделены и удалены с помощью метода drop_duplicates(). В результате количество строк уменьшилось с 16 679 до 16 444, то есть удалено 235 дублирующих записей.

In [28]:
#Ввиду того, что после проведения операций по обработке и очистке пропусков и дубликатов у нас автоматически изменились
#типы данных в стоолбцах `platform` и `rating`, произведем повторное преобразование в category
types_to_convert = ['platform',  'rating']
ng.loc[:, types_to_convert] = ng.loc[:, types_to_convert].astype('category')
ng.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16444 entries, 15191 to 9260
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype   
---  ------           --------------  -----   
 0   name             16444 non-null  object  
 1   platform         16444 non-null  object  
 2   year_of_release  16444 non-null  int64   
 3   genre            16444 non-null  category
 4   na_sales         16444 non-null  float64 
 5   eu_sales         16444 non-null  float64 
 6   jp_sales         16444 non-null  float64 
 7   other_sales      16444 non-null  float64 
 8   critic_score     16444 non-null  float64 
 9   user_score       16444 non-null  float64 
 10  rating           9768 non-null   object  
dtypes: category(1), float64(6), int64(1), object(3)
memory usage: 1.4+ MB


- В процессе подготовки данных мы обрабатывали и удаляли строки с пропусками, ошибками и дубликатами. В таком случае посчитаем количество удалённых строк в абсолютном и относительном значениях.

In [29]:
deleted_rows = initial_row_count1 - final_row_count2
deleted_percent_rows = (deleted_rows / initial_row_count1) * 100
display(f"Удалено строк: {deleted_rows}, что составляет: {deleted_percent_rows:.2f}%")

'Удалено строк: 512, что составляет: 3.02%'

На этапе обработки данных были предприняты следующие шаги:
  - Для столбцов `critic_score` и `user_score` пропуски были заменены на значение -1, что позволяет учитывать отсутствующие оценки при дальнейшем анализе. Дополнительно, были предприняты меры по удалению пропусков в столбцах `name`, `genre` и `year_of_release` с помощью функции dropna(). После этого пропуски в столбцах `eu_sales` и `jp_sales` были заполнены средними значениями по группам, сформированным на основе платформы и года выпуска игры.
  - Был проведен анализ уникальных значений и нормализация категориальных данных. В частности, для столбца `genre` все текстовые значения были приведены к нижнему регистру, что помогло устранить неявные дубликаты. Дополнительно было проверено содержание некорректных категорий в столбцах `platform` и `year_of_release` - дубликатов или ошибок обнаружено не было.
  - В процессе очистки были выявлены и удалены явные дубликаты. Общее число обнаруженных дубликатов составило 235. Повторяющиеся записи были устранены путём сортировки данных по всем столбцам и удаления с помощью функции drop_duplicates().
  
  В результате объём данных сократился с первоначальных 16956 до 16444 строк.
  

---

## 3. Фильтрация данных

Перед нами стоит цель изучить историю продаж игр в начале XXI века в период с 2000 по 2013 год включительно. Отберем данные по этому показателю. Сохраним новый срез данных в отдельном датафрейме `ng_xxi`.

In [30]:
#Произведем фильтрацию по столбцу `year_of_release` с помощью логической маски
ng_xxi = ng[(ng['year_of_release'] >= 2000) & (ng['year_of_release'] <= 2013)].copy()
display(ng_xxi.head())

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
3394,Frozen: Olaf's Quest,3DS,2013,platform,0.27,0.27,0.00,0.05,-1.0,-1.0,NaN
3906,Frozen: Olaf's Quest,DS,2013,platform,0.21,0.26,0.00,0.04,-1.0,-1.0,NaN
2478,Tales of Xillia 2,PS3,2012,role-playing,0.20,0.12,0.45,0.07,71.0,7.9,T
8460,.hack//G.U. Vol.1//Rebirth,PS2,2006,role-playing,0.00,0.00,0.17,0.00,-1.0,-1.0,NaN
7182,.hack//G.U. Vol.2//Reminisce,PS2,2006,role-playing,0.11,0.09,0.00,0.03,-1.0,-1.0,NaN


---

## 4. Категоризация данных
    
Проведем категоризацию данных:
- Разделим все игры по оценкам пользователей и выделим такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [31]:
#Категоризуем игры по оценкам пользователей
user_grade_groups = []

for grade in ng_xxi['user_score']:
    if 8 <= grade <= 10:
        user_grade_groups.append("высокая оценка")
    elif 3 <= grade < 8:
        user_grade_groups.append("средняя оценка")
    elif 0 <= grade < 3:
        user_grade_groups.append("низкая оценка")
    else:
        user_grade_groups.append("Нет оценки")
        
ng_xxi['user_grade_group'] = user_grade_groups

- Разделим все игры по оценкам критиков и выделим такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [32]:
#Категоризуем игры по оценкам критиков
critic_grade_groups = []

for grade in ng_xxi['critic_score']:
    if 80 <= grade <= 100:
        critic_grade_groups.append("высокая оценка")
    elif 30 <= grade < 80:
        critic_grade_groups.append("средняя оценка")
    elif 0 <= grade < 30:
        critic_grade_groups.append("низкая оценка")
    else:
        critic_grade_groups.append("Нет оценки")
        
ng_xxi['critic_grade_group'] = critic_grade_groups
display(ng_xxi.head())

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating,user_grade_group,critic_grade_group
3394,Frozen: Olaf's Quest,3DS,2013,platform,0.27,0.27,0.00,0.05,-1.0,-1.0,NaN,Нет оценки,Нет оценки
3906,Frozen: Olaf's Quest,DS,2013,platform,0.21,0.26,0.00,0.04,-1.0,-1.0,NaN,Нет оценки,Нет оценки
2478,Tales of Xillia 2,PS3,2012,role-playing,0.20,0.12,0.45,0.07,71.0,7.9,T,средняя оценка,средняя оценка
8460,.hack//G.U. Vol.1//Rebirth,PS2,2006,role-playing,0.00,0.00,0.17,0.00,-1.0,-1.0,NaN,Нет оценки,Нет оценки
7182,.hack//G.U. Vol.2//Reminisce,PS2,2006,role-playing,0.11,0.09,0.00,0.03,-1.0,-1.0,NaN,Нет оценки,Нет оценки


- После категоризации данных проверим результат: сгруппируем данные по выделенным категориям и посчитаем количество игр в каждой категории.

In [33]:
#Произведем группировку по столбцам `user_grade_group` и `critic_grade_group` и подсчитаем кол-во игр по выделенным 
#категориям
count_games_category = ng_xxi.groupby(['critic_grade_group', 'user_grade_group'])['name'].count()
display(count_games_category)

critic_grade_group  user_grade_group
Нет оценки          Нет оценки          5237
                    высокая оценка       101
                    низкая оценка         21
                    средняя оценка       253
высокая оценка      Нет оценки            33
                    высокая оценка      1017
                    низкая оценка          1
                    средняя оценка       641
низкая оценка       Нет оценки             7
                    высокая оценка         1
                    низкая оценка         17
                    средняя оценка        30
средняя оценка      Нет оценки          1021
                    высокая оценка      1167
                    низкая оценка         77
                    средняя оценка      3157
Name: name, dtype: int64

- Выделим топ-7 платформ по количеству игр, выпущенных за актуальный период с 2000 по 2013 год включительно.

In [34]:
#Произведем подсчет игр по каждой платформе и выделим 7 из них с наибольшим количеством
count_games_platform = ng_xxi['platform'].value_counts()
display(count_games_platform.head(7))

platform
PS2     2127
DS      2120
Wii     1275
PSP     1180
X360    1121
PS3     1087
GBA      811
Name: count, dtype: int64

## 5. Итоговый вывод

В рамках проекта была проведена комплексная подготовка данных, включающая загрузку и ознакомление с исходным датасетом, проверку ошибок и их предобработку: исправление названий и меток столбцов, приведение типов данных к нужному формату, обработку пропусков, а также выявление и устранение явных и неявных дубликатов.
Затем для категорий оценки критиков был создан новый столбец critic_grade_group, а для оценок пользователей - user_grade_group, что позволило более удобно сегментировать игры по уровням: «Нет оценки», «Низкая оценка», «Средняя оценка» и «Высокая оценка», выявив на основании проведенной групировки данных по этим столбцам и дальнейшем подсчете количества игр по категориям, что большинство из них имеют либо высокие, либо средние, либо вовсе не имеют оценки - например, игр с высокой оценкой критиков и пользователей более 1000, а игр без оценки - свыше 5000. 
Также дополнительно определили топ-7 платформ по количеству изданных игр за период с 2000 по 2013 год включительно: PS2 (2127 игр), DS (2120), Wii (1275), PSP (1180), X360 (1121), PS3 (1087) и GBA (811). Эти платформы доминируют на рынке в рассматриваемый период и охватывают большую часть базы игроков.

В результате проделанной работы мы получили чистый, структурированный и расширенный датасет, который позволяет выявлять ключевые тренды развития игровой индустрии, популярные платформы и жанры, а также оценки пользователей и критиков. Эта база станет основой для разработки стратегий привлечения аудитории и продвижения игры «Секреты Темнолесья» как среди фанатов классических RPG, так и новых игроков.
